### PR TITLE
[Staged Recipes Integration] Retire `arch_manager` in the transformer pipeline.

### PR DESCRIPTION
**Pull request by dbogunowicz**
This PR:
- removes `arch_managers` from the `transformers` pipeline.
- adds some minor modifications to the original PR it is derived from (#636)

Manual tests performed:
- verified the behavior of `qat_active` in L738 of `src/sparseml/transformers/sparsification/trainer.py` (by passing a  `self.arch_recipe` a recipe with or without quantization modifiers.)
- verified the behavior of `export_transformer_to_onnx` in L179-189 of `src/sparseml/transformers/export.py`
- verified the training pipeline E2E:
1. verified `_extract_metadata` method (L410 in `sparsification/trainer.py`
2. verified `_setup_manager` method (L432 in `sparsification/trainer.py` by making sure that both `self.recipe` and `arch_recipe` exist and are valid.
```
2022-04-06 10:01:43 sparseml.transformers.sparsification.trainer INFO     Loaded SparseML recipe variable into manager for recipe: zoo:nlp/masked_language_modeling/bert-base/pytorch/huggingface/wikipedia_bookcorpus/12layer_pruned80_quant-none-vnni?recipe_type=transfer-question_answering, recipe_variables: None and metadata OrderedDict([('per_device_train_batch_size', 16), ('learning_rate', 5e-05), ('num_train_epochs', 3.0), ('warmup_steps', 0), ('fp16', True)])
2022-04-06 10:01:58 sparseml.transformers.sparsification.trainer INFO     Loaded SparseML 1 recipe stage(s) into architecture managers from /home/damian/.cache/sparsezoo/03af0342-d7d0-470b-8958-80230ff0af10/pytorch/recipe.yaml
```
3. verified `apply_manager`
```
2022-04-06 10:42:02 sparseml.transformers.sparsification.trainer INFO     Applied structure from 1 previous recipe stage(s) to model and finalized (recipes saved with model_path)
2022-04-06 10:42:02 sparseml.transformers.sparsification.trainer INFO     Applied structure from SparseML recipe argument to model at epoch 0.0
```
4. verified `save` (both during training (saving checkpoints, as well as after training)

```
version: 1.1.0

__metadata__:

modifiers:
    - !EpochRangeModifier
        end_epoch: 8
        start_epoch: 0.0

    - !LearningRateFunctionModifier
        cycle_epochs: 1.0
        end_epoch: 8
        final_lr: 0
        init_lr: 0.00015
        lr_func: linear
        start_epoch: 0
        update_frequency: -1.0
```

plus 

```
version: 1.1.0

__metadata__:
  per_device_train_batch_size: 16
  learning_rate: 5e-05
  num_train_epochs: 3.0
  warmup_steps: 0
  fp16: True

modifiers:
    - !EpochRangeModifier
        end_epoch: 13
        start_epoch: 0.0

    - !LearningRateFunctionModifier
        cycle_epochs: 1.0
        end_epoch: 13
        final_lr: 0
        init_lr: 0.00015
        lr_func: linear
        start_epoch: 0
        update_frequency: -1.0

    - !DistillationModifier
        distill_output_keys: ['start_logits', 'end_logits']
        end_epoch: -1.0
        hardness: 1.0
        start_epoch: -1.0
        temperature: 2.0
        update_frequency: -1.0

    - !ConstantPruningModifier
        end_epoch: -1.0
        params: __ALL_PRUNABLE__
        start_epoch: 0.0
        update_frequency: -1

    - !QuantizationModifier
        disable_quantization_observer_epoch: 12.0
        end_epoch: -1.0
        exclude_module_types: ['LayerNorm', 'Tanh']
        freeze_bn_stats_epoch: 12.0
        quantize_embeddings: 1
        quantize_linear_activations: 0
        reduce_range: False
        start_epoch: 8.0
        submodules: ['qa_outputs', 'bert.embeddings', 'bert.encoder']
```

equals

```
version: 1.1.0

stage_0:
  __metadata__: None

  stage_0_modifiers:
      - !EpochRangeModifier
          end_epoch: 8
          start_epoch: 0.0
  
      - !LearningRateFunctionModifier
          cycle_epochs: 1.0
          end_epoch: 8
          final_lr: 0
          init_lr: 0.00015
          lr_func: linear
          start_epoch: 0
          update_frequency: -1.0
  

stage_1:
  __metadata__:
    per_device_train_batch_size: 16
    learning_rate: 5e-05
    num_train_epochs: 3.0
    warmup_steps: 0
    fp16: True

  stage_1_modifiers:
      - !DistillationModifier
          distill_output_keys: ['start_logits', 'end_logits']
          end_epoch: 7.0
          hardness: 1.0
          start_epoch: 7.0
          temperature: 2.0
          update_frequency: -1.0
  
      - !ConstantPruningModifier
          end_epoch: 7.0
          params: __ALL_PRUNABLE__
          start_epoch: 8.0
          update_frequency: -1
  
      - !QuantizationModifier
          disable_quantization_observer_epoch: 12.0
          end_epoch: 7.0
          exclude_module_types: ['LayerNorm', 'Tanh']
          freeze_bn_stats_epoch: 12.0
          quantize_embeddings: 1
          quantize_linear_activations: 0
          reduce_range: False
          start_epoch: 16.0
          submodules: ['qa_outputs', 'bert.embeddings', 'bert.encoder']
  
      - !EpochRangeModifier
          end_epoch: 21
          start_epoch: 8.0
  
      - !LearningRateFunctionModifier
          cycle_epochs: 1.0
          end_epoch: 21
          final_lr: 0
          init_lr: 0.00015
          lr_func: linear
          start_epoch: 8
          update_frequency: -1.0
```

### CODE REVIEW

File: src/sparseml/transformers/sparsification/trainer.py [hunk 1]
```diff
@@ -114,11 +108,17 @@ def __init__(
         self.recipe = recipe
         self.recipe_args = recipe_args
         self.teacher = teacher
+
+        # hack for now, maybe better to add training_args as
```
Comment (bfineran) [inline src/sparseml/transformers/sparsification/trainer.py:112]: why do we need to extract metadata for export mode?

File: src/sparseml/transformers/sparsification/trainer.py [hunk 2]
```diff
@@ -444,52 +421,29 @@ def _setup_manager(
...
     ) -> Tuple[Optional[ScheduledModifierManager], List[ScheduledModifierManager]]:
         manager = None
-        arch_managers = []
-
+        checkpoint_recipe = os.path.join(self.model_state_path, RECIPE_NAME)
+        # By default we should fetch checkpoint_recipe from self.model_state_path
+        # This is used for testing only
+        # checkpoint_recipe = os.path.join(
+        #    kwargs["args"].to_dict()["output_dir"], RECIPE_NAME
+        # )
```
Comment (bfineran) [inline src/sparseml/transformers/sparsification/trainer.py:429]: does this need to be here in prod?

File: src/sparseml/transformers/sparsification/trainer.py [hunk 3]
```diff
@@ -168,30 +167,16 @@ def apply_manager(self, epoch: float, checkpoint: Optional[str]) -> bool:
         :return: True if recipes were applied, False otherwise
         """
 
-        if (not self.arch_managers and self.manager is None) or self.manager_applied:
+        if self.manager is None or self.manager_applied:
             return False
 
         orig_state_dict = self.model.state_dict()
+        self.manager.apply_structure(self.model, epoch=epoch)
```
Comment (bfineran) [inline src/sparseml/transformers/sparsification/trainer.py:174]: we should also be applying the checkpoint recipe here, not the main recipe manager. what the arch managers were doing here was applying the previous recipes to reproduce graph state.

File: src/sparseml/transformers/sparsification/trainer.py [hunk 4]
```diff
@@ -140,8 +140,7 @@ def __init__(
         else:
             self.logger_manager = LoggerManager(log_python=False)
 
-        # remove arch_managers once recipe stages are supported
-        self.manager, self.arch_managers = self._setup_manager(kwargs)
+        self.manager = self._setup_manager(kwargs)
```
Comment (bfineran) [inline src/sparseml/transformers/sparsification/trainer.py:143]: should have checkpoint manager here as well still (you can think of all arch managers as being condensed into a single staged checkpoint recipe)

File: src/sparseml/transformers/sparsification/trainer.py [hunk 5]
```diff
@@ -444,52 +421,29 @@ def _setup_manager(
...
+            manager = ScheduledModifierManager.compose_staged(
+                base_recipe=checkpoint_recipe, additional_recipe=manager
+            )
             _LOGGER.info(
                 "Loaded SparseML recipe variable into manager for recipe: "
                 f"{self.recipe}, recipe_variables: {self.recipe_args} "
                 f"and metadata {self.metadata}"
             )
 
-        arch_recipe_paths = glob.glob(os.path.join(self.model_state_path, RECIPE_REGEX))
```
Comment (bfineran) [inline src/sparseml/transformers/sparsification/trainer.py:461]: I think we can get rid of RECIPE_REGEX and stuff around that now, but please double check

**Review by bfineran — CHANGES_REQUESTED** [review]

File: src/sparseml/optim/manager.py [hunk 6]
```diff
@@ -97,6 +97,16 @@ def __eq__(self, compare: object) -> bool:
     def metadata(self):
         return self._metadata
 
+    def number_stages(self) -> int:
```
Comment (bfineran) [inline src/sparseml/optim/manager.py:100]: nit: prefer `num_stages` to match repo naming conventions

File: src/sparseml/transformers/sparsification/trainer.py [hunk 7]
```diff
@@ -44,11 +42,7 @@
     WANDBLogger,
 )
 from sparseml.transformers.utils import SparseAutoModel
-from sparseml.transformers.utils.helpers import (
-    RECIPE_NAME,
-    RECIPE_REGEX,
-    RECIPE_TEMPLATE,
```
Comment (bfineran) [inline src/sparseml/transformers/sparsification/trainer.py:50]: are these args used anywhere now? if not, can delete them

File: src/sparseml/transformers/sparsification/trainer.py [hunk 8]
```diff
@@ -758,8 +736,13 @@ def check_disable(self, epoch: float, force: bool = False):
...
     def qat_active(self, epoch: float) -> bool:
-        return (self.trainer.manager and self.trainer.manager.qat_active(epoch)) or any(
-            bool(man.quantization_modifiers) for man in self.trainer.arch_managers
+        return (
+            self.trainer.manager
+            and self.trainer.manager.qat_active(epoch)
+            or any(
+                bool(stage.quantization_modifiers)
+                for stage in self.trainer.arch_manager.modifiers
+            )
```
Comment (bfineran) [inline src/sparseml/transformers/sparsification/trainer.py:745]:
```suggestion
            or self.trainer.arch_manager.quantization_modifiers
```

quantization_modifiers calls into `iter_modifiers()` under the hood which flattens stages for you!

**Review by bfineran — CHANGES_REQUESTED** [review]
looks good, few small suggestions.

Please make sure to do some testing and describe what testing was done in PR description

File: src/sparseml/transformers/export.py [hunk 9]
```diff
@@ -175,8 +175,18 @@ def export_transformer_to_onnx(
...
-        total_recipes = (1 if trainer.manager else 0) + len(trainer.arch_managers)
-        _LOGGER.info(f"Applied {total_recipes} total recipes the model at {model_path}")
+        num_stages = 0
+        if trainer.manager:
+            num_stages += trainer.manager.num_stages()
+        if trainer.arch_manager:
+            num_stages += trainer.arch_manager.num_stages()
+
+        msg = (
+            "an unstaged_recipe"
```
Comment (dbogunowicz) [inline src/sparseml/transformers/export.py:185]:
```suggestion
            "an unstaged recipe"
```

**Review by bfineran — CHANGES_REQUESTED** [review]
@dbogunowicz update and tests look good, ready to approve, but looks like branch is in a bad state, can you make sure to clean up the extra changes then can approve

File: src/sparseml/transformers/sparsification/trainer.py [hunk 10]
```diff
@@ -444,38 +433,27 @@ def _setup_manager(
...
-            # remove this logic once recipe stages land
-            arch_managers.remove(manager)
+        arch_recipe = os.path.join(self.model_state_path, RECIPE_NAME)
+        if os.path.isfile(arch_recipe):
+            arch_manager = ScheduledModifierManager.from_yaml(arch_recipe)
             _LOGGER.info(
-                "Removed duplicate SparseML recipe from arch_managers that matched "
-                "the recipe variable to prevent double application"
+                f"Loaded SparseML {arch_manager.num_stages()} recipe stage(s) "
+                f"into architecture managers from {arch_recipe}"
```
Comment (bfineran) [inline src/sparseml/transformers/sparsification/trainer.py:456]:
```suggestion
                f"Loaded {arch_manager.num_stages()} SparseML checkpoint recipe stage(s) "
                f"from {arch_recipe} to replicate model sparse state"
```

File: src/sparseml/transformers/sparsification/trainer.py [hunk 11]
```diff
@@ -444,38 +433,27 @@ def _setup_manager(
...
-            # remove this logic once recipe stages land
-            arch_managers.remove(manager)
+        arch_recipe = os.path.join(self.model_state_path, RECIPE_NAME)
+        if os.path.isfile(arch_recipe):
+            arch_manager = ScheduledModifierManager.from_yaml(arch_recipe)
             _LOGGER.info(
-                "Removed duplicate SparseML recipe from arch_managers that matched "
-                "the recipe variable to prevent double application"
+                f"Loaded {arch_manager.num_stages()} SparseML checkpoint recipe stage(s) "
+                f"from {arch_recipe} to replicate model sparse state"
```
Comment (bfineran) [inline src/sparseml/transformers/sparsification/trainer.py:456]:
```suggestion
                f"Loaded {arch_manager.num_stages()} SparseML checkpoint recipe "
                f"stage(s) from {arch_recipe} to replicate model sparse state"
```

**Review by bfineran — DISMISSED** [review]
LGTM

File: src/sparseml/transformers/sparsification/trainer.py [hunk 12]
```diff
@@ -758,8 +736,10 @@ def check_disable(self, epoch: float, force: bool = False):
             self.disable_amp(epoch)
 
     def qat_active(self, epoch: float) -> bool:
-        return (self.trainer.manager and self.trainer.manager.qat_active(epoch)) or any(
-            bool(man.quantization_modifiers) for man in self.trainer.arch_managers
+        return (
+            self.trainer.manager
+            and self.trainer.manager.qat_active(epoch)
+            or self.trainer.arch_manager.quantization_modifiers
```
Comment (KSGulin) [inline src/sparseml/transformers/sparsification/trainer.py:742]: There may be a potential error case here. If `self.trainer.manager.qat_active(epoch)` is false and `arch_manager` is None, then trying to access self.trainer.arch_manager.quantization_modifiers will throw an error. Is such a case possible?

Comment (dbogunowicz) [reply src/sparseml/transformers/sparsification/trainer.py:742]: good corner-case hunting! testing

**Review by bfineran — APPROVED** [review]

**Review by KSGulin — APPROVED** [review]

**Pull request merged**